File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

const BACKSTAGE_PASSES = "Backstage passes to a TAFKAL80ETC concert";
const SULFURAS = "Sulfuras, Hand of Ragnaros";
const AGED_BRIE = "Aged Brie";

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    this.items.forEach((item) => {
      this.updateItemQuality(item);
    });

    return this.items;
  }

  private updateItemQuality(item: Item) {
    if (item.name != AGED_BRIE && item.name != BACKSTAGE_PASSES) {
      if (item.quality > 0) {
        if (item.name != SULFURAS) {
          item.quality = item.quality - 1;
        }
      }
    } else {
      if (item.quality < 50) {
        item.quality = item.quality + 1;
        if (item.name == BACKSTAGE_PASSES) {
          if (item.sellIn < 11) {
            if (item.quality < 50) {
              item.quality = item.quality + 1;
            }
          }
          if (item.sellIn < 6) {
            if (item.quality < 50) {
              item.quality = item.quality + 1;
            }
          }
        }
      }
    }
    if (item.name != SULFURAS) {
      item.sellIn = item.sellIn - 1;
    }
    if (item.sellIn < 0) {
      if (item.name != AGED_BRIE) {
        if (item.name != BACKSTAGE_PASSES) {
          if (item.quality > 0) {
            if (item.name != SULFURAS) {
              item.quality = item.quality - 1;
            }
          }
        } else {
          item.quality = item.quality - item.quality;
        }
      } else {
        if (item.quality < 50) {
          item.quality = item.quality + 1;
        }
      }
    }
  }
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            if item.name == "Sulfuras, Hand of Ragnaros":
                return
            self.reduce_sell_in(item)

            if item.name == "Aged Brie":
                if item.quality < 50:
                    self.increase_item_quality(item)
                return

            if item.name == "Backstage passes to a TAFKAL80ETC concert":
                if item.quality < 50:
                    self.increase_item_quality(item)
                if item.sell_in < 11:
                    self.increase_item_quality(item)
                if item.sell_in < 6:
                    self.increase_item_quality(item)
                if item.sell_in < 0:
                    item.quality = item.quality - item.quality
                return

            if item.quality > 0:
                self.decrease_item_quality(item)

    def reduce_sell_in(self, item):
        item.sell_in -= 1

    def increase_item_quality(self, item):
        item.quality += 1

    def decrease_item_quality(self, item):
        item.quality -= 1


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

    def reduce_sell_in(self):
        self.sell_in -= 1

    def increase_item_quality(self):
        self.quality += 1

    def decrease_item_quality(self):
        self.quality -= 1


class Sulfuras(Item):
    pass


class AgedBrie(Item):
    pass


class BackstagePasses(Item):
    pass


class ItemFactory:
    def create_item(name: str, sell_in: int, quality: int):
        if name == "Sulfuras, Hand of Ragnaros":
            return Sulfuras(name, sell_in, quality)
        if name == "Aged Brie":
            return AgedBrie(name, sell_in, quality)
        if name == "Backstage passes to a TAFKAL80ETC concert":
            return BackstagePasses(name, sell_in, quality)
        else:
            return Item(name, sell_in, quality)
